chore: activate noImplicitAny

Shinigami92 opened the pull request:
<!-- Please first read https://github.com/faker-js/faker/blob/verify-semantic-pull-requests/CONTRIBUTING.md -->

<!-- Help us by writing a correct PR title following this guide: https://github.com/faker-js/faker/blob/verify-semantic-pull-requests/CONTRIBUTING.md#committing -->

closes #1837

This PR does only build time and no runtime changes on purpose
Otherwise I would suggest to modify some codeflow in later PRs (if really needed)

Shinigami92 committed:
chore: activate noImplicitAny

Shinigami92 committed:
type module constructor hacks

Shinigami92 committed:
type variadic args

Shinigami92 committed:
fix other stuff

xDivisionByZerox commented on src/modules/helpers/unique.ts:
```diff
@@ -81,7 +81,7 @@ Try adjusting maxTime or maxRetries parameters for faker.helpers.unique().`
  * @param options.store The store of unique entries. Defaults to `GLOBAL_UNIQUE_STORE`.
  */
-export function exec<Method extends (...parameters) => RecordKey>(
+export function exec<Method extends (...parameters: any[]) => RecordKey>(
```
<s>This can be expressed in a more generic way, to avoid explicit any (same in index.ts):

```suggestion
export function exec<Method extends <T>(...parameters: T[]) => RecordKey>(
```

@ST-DDT, can you confirm this? I'm "only" 90% sure that it works as expected. </s>

xDivisionByZerox replied:
NVM, I had massive brain lag or something... this doesn't work as we require the parameters generically as well. MB.

xDivisionByZerox commented on src/utils/types.ts:
```diff
@@ -12,7 +12,7 @@ export type LiteralUnion<T extends U, U = string> =
  * For some strange reason this is not the same as `Function`.
  */
-export type Callable = (...args) => unknown;
+export type Callable = (...args: any[]) => unknown;
```
<s>This can be typed generically to avoid explicit `any`:

```suggestion
export type Callable = <T>(...args: T[]) => unknown;
```
</s>

NVM, doesn't work when you have multiple parameters with different types.
BUT: Since we are only interested in whether a type is a function (callable), this type can be narrowed to `export type Callable = () => void;`, since all functions match this signature.

```suggestion
export type Callable = () => void;
```

The description might need an update as well.

Shinigami92 replied:
I'm getting this

<img width="800" alt="image" src="https://user-images.githubusercontent.com/7195563/218706286-8239065d-7a17-42e0-a468-5d98f1446284.png">

Please help me further
I suggest checkout the branch and try to handle it on your own with your knowledge

xDivisionByZerox replied:
You are too quick xD. I updated my previous comment.

Shinigami92 replied:
Okay I mark these as resolved

xDivisionByZerox replied:
Please read my suggestion again. Guess you only read the first line.

> Since we are only interested in whether a type is a function (callable), this type can be narrowed to `export type Callable = () => void;`

Shinigami92 committed:
Merge branch 'next' into 1837-turn-on-noimplicitany-in-tsconfig

Shinigami92 committed:
Inline () => void

Shinigami92 commented:
@xDivisionByZerox Oh I see, the Callable was just a replacement for `Function` which should not be used in TS as a type as it is a real thing (e.g. like String)

The suggested way is just to use `() => void` instead and therefore I removed the type alias and inlined it as this is the way it should be done and is suggested by TS

This also removed the _bad worded_ comment `For some strange reason ...` which sounds a bit unprofessional

Shinigami92 commented:
The "test"-code failed 👀 
I need to check that later, I will leave my home soon and come back in some hours

ST-DDT commented on src/modules/helpers/index.ts:
```diff
@@ -903,7 +907,7 @@ export class HelpersModule {
    * @since 7.5.0
    */
-  unique<Method extends (...parameters) => RecordKey>(
+  unique<Method extends (...parameters: any[]) => RecordKey>(
```
This has a lint warning.

Shinigami92 replied:
Yes, this is okay for this PR and should be addressed in a separate PR if needed
This PR just aims first to toggle on this TypeScript flag and is not related to eslint

ST-DDT replied:
At least mark them as ignored, otherwise they will be shown in Github in every PR as warnings.

ST-DDT commented on src/modules/helpers/unique.ts:
```diff
@@ -81,7 +81,7 @@ Try adjusting maxTime or maxRetries parameters for faker.helpers.unique().`
  * @param options.store The store of unique entries. Defaults to `GLOBAL_UNIQUE_STORE`.
  */
-export function exec<Method extends (...parameters) => RecordKey>(
+export function exec<Method extends (...parameters: any[]) => RecordKey>(
```
Same here

ST-DDT commented on src/utils/types.ts:
```diff
@@ -7,17 +7,10 @@ export type LiteralUnion<T extends U, U = string> =
  * Type that represents a single method/function name of the given type.
  */
-export type MethodOf<ObjectType, Signature extends Callable = Callable> = {
+export type MethodOf<ObjectType, Signature extends () => void = () => void> = {
```
Maybe

```suggestion
export type MethodOf<ObjectType, Signature extends () => unknown = () => unknown> = {
```

Not tested

ST-DDT commented on src/utils/types.ts:
```diff
@@ -30,5 +23,5 @@ export type MethodOf<ObjectType, Signature extends Callable = Callable> = {
 export type MethodsOf<
   ObjectType,
-  Signature extends Callable = Callable
+  Signature extends () => void = () => void
```
Same here

ST-DDT commented:
> The suggested way is just to use `() => void` instead and therefore I removed the type alias and inlined it as this is the way it should be done and is suggested by TS
> 
> This also removed the _bad worded_ comment `For some strange reason ...` which sounds a bit unprofessional

Soon: ![](https://pbs.twimg.com/media/D9JjSQHXoAANj9J.jpg)

Shinigami92 committed:
bring back callback

Shinigami92 committed:
add todos

Shinigami92 committed:
Merge branch 'next' into 1837-turn-on-noimplicitany-in-tsconfig

xDivisionByZerox commented on src/utils/types.ts:
```diff
@@ -8,11 +8,15 @@ export type LiteralUnion<T extends U, U = string> =
+ * A function that returns a value.
  *
- * For some strange reason this is not the same as `Function`.
+ * This is a workaround for the fact that `Function` is a real JS Object like `String` and therefore should not be used as a type.
```
That's not the correct for why the linter doesn't allow it. The `Function` type satisfies any function. This includes classes. So while this is correct from a type perspective it not desired most of the time:

```ts
function foo(fn: Function) {
  return fn();
}

foo(() => null); // this is fine and desired

function bar() {  
  return null; 
}
foo(); // this is fine and desired

class A { }
foo(A); // this is fine for typescript but not desired as `A` will be called without the `new` keyword which throws an error at runtime
```

---

Just wanted to point that out, since it has nothing to do with the actual type values being used as a type.

Shinigami92 replied:
Would you suggest something different?

Here might be some context for improved text: https://typescript-eslint.io/rules/ban-types

xDivisionByZerox replied:
How about this?

```suggestion
 * `Function` cannot be used instead of this, as it doesn't accept accepts class declarations.
 * Those would fail when called, as they are called without the `new` keyword.
```

Shinigami92 committed:
Update src/utils/types.ts

Co-authored-by: DivisionByZero <leyla.jaehnig@gmx.de>

Shinigami92 commented on src/utils/types.ts:
```diff
@@ -8,11 +8,16 @@ export type LiteralUnion<T extends U, U = string> =
+ * A function that returns a value.
  *
- * For some strange reason this is not the same as `Function`.
+ * `Function` cannot be used instead of this, as it doesn't accept accepts class declarations.
+ * Those would fail when called, as they are called without the `new` keyword.
```
```suggestion
 * `Function` cannot be used instead because it doesn't accept class declarations.
 * These would fail when invoked since they are invoked without the `new` keyword.
```

Shinigami92 committed:
Update src/utils/types.ts

ST-DDT committed:
Merge branch 'next' into 1837-turn-on-noimplicitany-in-tsconfig
